### PR TITLE
speed up reading from control socket

### PR DESCRIPTION
**Pull request by mkomon**
This patch solves issue #5.

The overhead of appending chunks to data buffer in each pass of the loop is too high and progressively slows down the execution with each pass. This is a problem for large amounts of data such as full BGP table. This overhead must be avoided and the chunks of data must be stored as they are read and joined only once at the very end.

This speeds up the reading of large output by more than 500x (what would not complete in 30 minutes now takes less than 5 seconds).

All tests are passing.

### CODE REVIEW

File: pybird/__init__.py [hunk 1]
```diff
@@ -693,20 +693,21 @@ def _socket_query(self, query):
...
+            else:
+                tail = b''.join(data[-2:])[-256:].decode('utf-8')
+            if any([tail.find(f"\n{code:04}") != -1
                 for code in self.error_fields + self.success_fields]):
                 break
-            this_read = sock.recv(1024)
-            if not this_read:
-                self.log.debug(data)
-                raise ValueError("Could not read additional data from BIRD")
-            data += this_read
```
Comment (mkomon) [inline pybird/__init__.py:706]: this kills the performance and results it a fully utilized CPU core that keeps copying the buffer over and over

**Comment by grizz** [comment]
Looks good to me, thanks for the fix.

I'm adding the CI tests etc and we will do a release.

**Comment by grizz** [comment]
@mkomon the tests need a little tweaking, would you mind emailing me to discuss?

**Comment by mkomon** [comment]
To add some numbers about the speed-up factor, reading the current full IPv6 table of 161786 routes takes 3.5 hours before this patch (on my reasonably sized VPS) and it's down to less than 9 seconds (including parsing) with the patch, which would be 1400x speedup (in reality most of the 9 seconds is spent on parsing, so the difference is even greater).

The reading of the full IPv4 table of +900k routes does not complete in 48 hours so the difference there is by the factor of 20,000x or more.

The takeaway is that this kind of growing a buffer is crazy ineffective for large amounts of data.

**Pull request merged**